spec.md

```diff
@@ -2734,6 +2780,270 @@ message NodeExpandVolumeResponse {
 | Unsupported capacity_range | 11 OUT_OF_RANGE | Indicates that the capacity range is not allowed by the Plugin. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST fix the capacity range before retrying. |
 
+### Group Controller Service RPCs
+
+#### `GroupControllerGetCapabilities`
+
+A GroupController Plugin MUST implement this RPC call. This RPC allows the CO to check the supported capabilities of group controller service provided by the Plugin.
+
+```protobuf
+message GroupControllerGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+
+message GroupControllerGetCapabilitiesResponse {
+  // All the capabilities that the group controller service supports.
+  // This field is OPTIONAL.
+  repeated GroupControllerServiceCapability capabilities = 1;
+}
+
+// Specifies a capability of the group controller service.
+message GroupControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+
+      // Indicates that the group controller plugin supports
+      // creating, deleting, and getting details of a volume
+      // group snapshot.
+      CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT = 1
+      [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+```
+
+##### GroupControllerGetCapabilities Errors
+
+If the plugin is unable to complete the GroupControllerGetCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+
+#### `CreateVolumeGroupSnapshot`
+
+**ALPHA FEATURE**
+
+A Group Controller Plugin MUST implement this RPC call if it has `CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT` controller capability.
+This RPC will be called by the CO to create a new volume group snapshot from a list of source volumes on behalf of a user.
+
+The purpose of this call is to request the creation of a multi-volume snapshot. Group snapshots can be created from the existing list of volumes. Note that calls to this function MUST be idempotent - the function may be called multiple times for the same name - the group snapshot must only be created once.
+
+If a group snapshot corresponding to the specified group snapshot `name` is successfully created (meaning all snapshots associated with the group are successfully cut), the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeGroupSnapshotResponse`.
+
+If an error occurs before a group snapshot is cut, `CreateVolumeGroupSnapshot` SHOULD return a corresponding gRPC error code that reflects the error condition.
+
+`CreateVolumeGroupSnapshot` SHOULD return `0 OK` after all the snapshots are cut regardless of whether post processing such as uploading is complete or not.
```

Comment:
We had a lengthy discussion regarding this: https://github.com/container-storage-interface/spec/pull/519#discussion_r1068918404

Yes, post-processing could fail. We have decided to drop the `ready_to_use` field from VolumeGroupSnapshot message in CSI spec after the above discussions. CO will check `ready_to_use` field of each individual snapshot.  Only after `ready_to_use` becomes true for all snapshots, CO will treat the group snapshot as ready_to_use.  In K8s VolumeGroupSnapshot API object, there will be a readyToUse field so a user can check that and know whether post-processing is complete for the entire group snapshot.

Reply:
I read the lengthy discussion (thanks), my question remains: if (optional) post-processing fails w/ an error code/message, how/where is that relayed to the user? i suppose this question applies to snapshots in general, not just group snapshots.

Reply:
For individual snapshot, we have detailed description in CSI spec:

https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot
```
For plugins that supports snapshot post processing such as uploading, CreateSnapshot SHOULD return 0 OK and ready_to_use SHOULD be set to false after the snapshot is cut but still being processed. CO SHOULD then reissue the same CreateSnapshotRequest periodically until boolean ready_to_use flips to true indicating the snapshot has been "processed" and is ready to use to create new volumes. If an error occurs during the process, CreateSnapshot SHOULD return a corresponding gRPC error code that reflects the error condition.
```

Reply:
For group snapshot, if an error occurs during the post-processing, CreateVolumeGroupSnapshot SHOULD also return a corresponding gRPC error code that reflects the error condition.  I'll add this.

Reply:
I'm thinking maybe we still need to have ready_to_use in VolumeGroupSnapshot.

Without ready_to_use in VolumeGroupSnapshot, the steps are as follows:
- CreateVolumeGroupSnapshot returns 0 after group snapshot is cut
- Post-processing ......
- CreateVolumeGroupSnapshot returns error code if post-processing from any one snapshot fails

CreateVolumeGroupSnapshot already finished successfully in step one, but then it failed.
We can't get error code from individual snapshot because CreateSnapshot is not called here. The only thing we can get from Snapshot message is the ready_to_use boolean field.

With ready_to_use in VolumeGroupSnapshot, the steps are as follows:
- CreateVolumeGroupSnapshot returns 0 and ready_to_use to false after group snapshot is cut
- Post-processing ......
- CreateVolumeGroupSnapshot returns error code and ready_to_use is still false if post-processing fails

Reply:
My thinking is that the CreateVolumeGroupSnapshot RPC's only job is to get the individual snapshots cut and return their IDs. At this point the CO is able to populate the snapshotHandle field of all the needed VSC objects and watch them for ready_to_use. Any errors past that point are going to be related to individual snapshots, and if the CO wants to handle those errors by cleaning up and trying again, it's going to have to call the DeleteVolumeGroupSnapshot API with the appropriate IDs regaradless of what changes we make to the create API. But that's up to the CO -- it could also choose to leave the partial group snapshot there and only use the successful subset and that's technically valid. The use case could be that snapshots are taken frequently enough that failures are ignored and old snapshots are simply reaped on a schedule.

Post-processing _of the group snapshot itself_ is only relevant when there is an API to clone the whole group snapshot to a new volume group, and I still feel like it can be added in the future as part of that feature (if we decide to go that route).

Reply:
I think we still need the wording there to provide guidance to CO.  In K8s VolumeGroupSnapshot API object, we still have readyToUse field so that an user can take a look of that one field and know whether the whole thing is complete or not.

Reply:
We could consider removing the field there too. If we don't, we just have the modify the behavior of the sidecar to set that field based on the logical AND of all the member snapshots' readyToUse fields.

The point is that we should define the CreateVolumeGroupSnapshot RPC to have the very simple job of creating a list of snapshots with the special guarantee that they were cut at the same time, and nothing else. This minimal guarantee allows many underlying implementations to all work well behind this interface.

Reply:
Chatted with @bswartz about this.  Added ready_to_use back to VolumeGroupSnapshot so that the behavior of CreateVolumeGroupSnapshot is consistent with CreateSnapshot.
https://github.com/container-storage-interface/spec/pull/519/commits/11d9ad36d464ebab9ef161134ec5096fb8ce1077